torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1115,57 +1162,71 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
+  });
 
-  workMetaListCV_.notify_one();
+  waitForFutureOrTimeout(
+      fut, kProcessGroupNCCLAbortTimeout, "ProcessGroup abort");
+  LOG(INFO) << logPrefix() << "ProcessGroupNCCL aborts successfully.";
+
+  // We need to wait for abort to finish before we can safely shut down
+  // heartbeat monitoring thread.
   terminateHeartbeatMonitorThread_.store(true);
   monitorWakeUpCV_.notify_one();
 }
 
 ProcessGroupNCCL::~ProcessGroupNCCL() {
   LOG(INFO) << logPrefix() << "ProcessGroupNCCL destructor entered.";
-  terminateProcessGroup_.store(true);
-  workMetaListCV_.notify_one();
 
+  if (!terminateProcessGroup_.load()) {
+    // Only if TORCH_NCCL_ABORT_IN_DESTROY_PG is enabled, terminateProcessGroup_
+    // will be set to true through destroy_process_group
+    if (abortInDestroyProcessGroup_) {
+      LOG(WARNING) << c10::str(
+          "WARNING: process group has NOT been destroyed before it is being destructed. ",
+          "On normal program exit, the application should call destroy_process_group to ",
+          "ensure that any pending NCCL data transfers have finished in this process. "
+          "In rare cases this process can exit before this point and block the progress of "
+          "another member of the process group. This constraint has always been present, "
+          " but this warning has only been added since PyTorch 2.3");
+    }
+    // If user haven't explicitly destroy/shutdown process group, destructor
+    // needs to do so
+    shutdown();
+  }
+
+  // Wait for all threads to finish before returning
 #ifdef ENABLE_NCCL_ERROR_CHECKING
   if (ncclCommWatchdogThread_.joinable()) {
     ncclCommWatchdogThread_.join();
+    LOG(INFO) << logPrefix() << "ProcessGroupNCCL watchdog thread joined.";
   }
-  LOG(INFO) << logPrefix() << "ProcessGroupNCCL watchdog thread joined.";
-#endif
-
-  if (onCompletionHookThread_.joinable())
-    onCompletionHookThread_.join();
-
-  // Abort communicators after all threads have exited to avoid having the
-  // threads dying due to aborted communicator and raising a SIGABRT
-  // We need to include PG information in the abort reason so we can tell the
-  // abort order.
-  std::string abortReason = c10::str("Process Group destroyed on rank ", rank_);
-  LOG(INFO)
-      << logPrefix()
-      << "ProcessGroupNCCL aborting communicators, check for 'abort finished' logs or look for abort hang";
-  abort(abortReason);
-  LOG(INFO) << logPrefix() << "ProcessGroupNCCL abort finished.";
-
-  // We need to wait for abort to finish before we can safely shut down
-  // heartbeat monitoring thread.
-  terminateHeartbeatMonitorThread_.store(true);
-  monitorWakeUpCV_.notify_one();
-#ifdef ENABLE_NCCL_ERROR_CHECKING
   if (ncclHeartbeatMonitorThread_.joinable()) {
     ncclHeartbeatMonitorThread_.join();
+    LOG(INFO) << logPrefix()
+              << "ProcessGroupNCCL heart beat monitor thread joined.";
   }
 #endif
+  if (onCompletionHookThread_.joinable()) {
```

Comment:
oh we never joined this thread before eh?

well thats shoddy. one instinct is to add that join in a separate PR just in case we have to revert it and can keep the rest of this code landed

Reply:
We called this join before, check line 1146 of this version of PR 
<img width="442" alt="image" src="https://github.com/pytorch/pytorch/assets/4711970/b3d4b243-e385-42bb-894b-f1ed39c99600">